specifyweb/frontend/js_src/lib/components/Formatters/Formatter.tsx

```diff
@@ -74,40 +102,64 @@ function Definitions({
   readonly item: GetSet<Formatter>;
 }): JSX.Element | null {
   const isReadOnly = React.useContext(ReadOnlyContext);
-  const handleChange = (fields: Formatter['definition']['fields']): void =>
-    setFormatter({
-      ...formatter,
-      definition: {
-        ...formatter.definition,
-        fields,
-      },
-    });
+
+  const handleChange = React.useCallback(
+    (fields: Formatter['definition']['fields']): void => {
+      setFormatter({
+        ...formatter,
+        definition: {
+          ...formatter.definition,
+          fields,
+        },
+      });
+    },
+    [formatter, setFormatter]
+  );
+
   const hasCondition = formatter.definition.conditionField !== undefined;
   const trimmedFields = hasCondition
     ? formatter.definition.fields
     : formatter.definition.fields.slice(0, 1);
 
   const table = formatter.table;
+
+  const handleChanged = (
+    field: Formatter['definition']['fields'][number],
+    index: number
+  ): void =>
+    handleChange(replaceItem(formatter.definition.fields, index, field));
+
+  React.useEffect(() => {
+    if (formatter.definition.fields.length === 0) {
+      handleChange([
+        ...formatter.definition.fields,
+        {
+          value: undefined,
+          fields: [],
+        },
+      ]);
+    }
+  }, [formatter.definition.fields, handleChange]);
```

Comment:
(optional micro-optimization):

```suggestion
  const needFormatter = formatter.definition.fields.length === 0;
  React.useEffect(() => {
    if (needFormatter) {
      handleChange([
        ...formatter.definition.fields,
        {
          value: undefined,
          fields: [],
        },
      ]);
    }
  }, [needFormatter, handleChange]);
```

This way the useEffect is only called when there is need to, rather than being called on every keystroke if user is typing.

don't think it would make any noticeable performance difference in this case, but in places like the query builder or workbench it might - so don't need to change anything here, just keep this in mind for the future